lib/rust/frp/src/fan.rs

```diff
@@ -0,0 +1,132 @@
+//! A utility with a single input allowing emitting values of different types and multiple
+//! outputs, one per type.
+
+use crate::prelude::*;
+
+use crate::AnyData;
+
+use std::any::TypeId;
+
+
+
+// ===========
+// === Fan ===
+// ===========
+
+struct FanOutput {
+    stream: AnyData,
+    runner: Rc<dyn Fn(&AnyData, &AnyData)>,
+}
+
+impl Debug for FanOutput {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "FanOutput")
+    }
+}
+
+/// A utility with a single input allowing emitting values of different types and multiple
+/// outputs, one per type. See tests to learn how to use it.
+#[allow(missing_docs)]
+#[derive(Debug)]
+pub struct Fan {
+    pub source: crate::Any<AnyData>,
+    map:        Map,
+}
+
+type Map = Rc<RefCell<HashMap<TypeId, FanOutput>>>;
+
+impl Fan {
+    /// Constructor.
+    pub fn new(network: &crate::Network) -> Self {
+        let map: Map = default();
+        crate::extend! { network
+            source <- any_mut::<AnyData>();
+            eval source([map] (event) {
+                if let Some(fan_output) = map.borrow_mut().get(&(*event.data).type_id()) {
+                    (fan_output.runner)(&fan_output.stream, event);
+                }
+            });
+        }
+        Self { source, map }
+    }
+
+    /// Get the output stream of the fan for the given type. See docs of [`Fan`] for an example.
+    ///
+    /// # Safety
+    /// This implementation is safe because we remember the [`TypeId`] when erasing the type, and we
+    /// query the right function based on the [`TypeId`] when emitting the value. When changing this
+    /// implementation, please remember to test it rigorously.
+    #[allow(unsafe_code)]
+    pub fn output<T: crate::Data>(&self, network: &crate::Network) -> crate::Source<T> {
+        let id = TypeId::of::<T>();
+        let mut map = self.map.borrow_mut();
+        let fan_output = map.entry(id).or_insert_with(|| {
+            crate::extend! { network
+                output <- source::<T>();
+            }
+            let stream = AnyData::new(output);
+            let runner = Rc::new(|stream: &AnyData, event: &AnyData| {
+                let stream = unsafe { stream.downcast_ref_unchecked::<crate::Source<T>>() };
+                let event = unsafe { event.downcast_ref_unchecked::<T>() };
+                stream.emit(event.clone());
```

Comment:
As we said during the call, in debug builds, make a checked downcast.

But honestly, even in release builds, I think this additional if is nothing compared to many allocations we make when an event is emitted.